lib/config_skeleton.rb

```diff
@@ -209,14 +222,24 @@ def run
     @terminate_r, @terminate_w = IO.pipe
 
     loop do
-      if ios = IO.select([notifier.to_io, @terminate_r], [], [], sleep_duration.tap { |d| logger.debug(logloc) { "Sleeping for #{d} seconds" } })
+      if ios = IO.select(
+          [notifier.to_io, @terminate_r, @trigger_regen_r],
+          [], [],
+          sleep_duration.tap { |d| logger.debug(logloc) { "Sleeping for #{d} seconds" } }
+      )
         if ios.first.include?(notifier.to_io)
           logger.debug(logloc) { "inotify triggered" }
           notifier.process
           regenerate_config(force_reload: true)
         elsif ios.first.include?(@terminate_r)
           logger.debug(logloc) { "triggered by termination pipe" }
           break
+        elsif ios.first.include?(@trigger_regen_r)
+          # check if anything was written, to exclude empty strings
+          if @trigger_regen_r.read_nonblock(1)
```

Comment:
we got to keep reading non blocking till we get an exception, otherwise over time we can get a backlog of data in the pipe.

Reply:
Why do we care about a backlog of data in the pipe? I only have this here because without it the `@trigger_regen_r` was making IO.select fire even though nothing was written yet. I don't need the text itself, just need to know that something is there. In your gist yesterday it is doing a `puts` on the data:

![image](https://user-images.githubusercontent.com/920448/100051127-d9c46200-2e66-11eb-9138-2c3816f88378.png)

Reply:
we care about backlog cause if stuff backs up enough we can run out of memory with buffers ... I think the correct pattern here is: 

```
while @trigger_regen_r.read_nonblock(20, nil, exception: false) != :wait_readable
  # do nothing
end
```

This makes sure the pipe is squeaky clean prior to proceeding with the next iteration.

Reply:
Ah I see, thank you for that. I should have read the `read_nonblock` docs closer...